File: cpp/newWave/15686.cpp
#include<bits/stdc++.h>

using namespace std;


int n ,m;
int arr[51][51];
vector<pair<int,int> > house;
vector<pair<int,int> > chicken;
int ans = 9999999;
int ch[14];
void go(int idx,int cnt){
	if(idx > chicken.size()){
		return;
	}
	if(cnt ==m){
		int s=0;
		for(int i=0 ; i<house.size(); i++){
			int dis=9999999;
			int yy=house[i].first;
			int xx = house[i].second;
			for(int j=0 ; j < chicken.size(); j++){
				if(ch[j]){
					dis =min(dis,abs(yy-chicken[j].first) + abs(xx-chicken[j].second));
				}
			}
			s+=dis;
		}
		ans= min(s,ans);
		return;
	}
	
	ch[idx] =1;
	go(idx+1, cnt+1);
	ch[idx]=0;
	go(idx+1,cnt);
}



int main(){
	
	cin>>n>>m;
	for(int i=0; i<n ;i++){
		for(int j=0; j<n; j++){
			cin>>arr[i][j];
			if(arr[i][j] ==1){
				house.push_back({i,j});
				
			}
			else if(arr[i][j] ==2){
				chicken.push_back({i,j});
			}
		}
	}
	
	
	
	
	go(0,0);
	
	
	cout<<ans<<endl;
	
	
	
	
	
	return 0;
}

File: python/15686.py
import sys

chicken = []
home = []
ans = sys.maxsize
ch = []


def go(idx, cnt):
    global ans
    if idx > len(chicken):
        return

    if cnt == M:
        sum = 0
        for i in range(len(home)):
            dis = sys.maxsize
            for j in range(len(chicken)):
                if ch[j] == 1:
                    dis = min(dis, abs(home[i][0] - chicken[j]
                                       [0]) + abs(home[i][1] - chicken[j][1]))
            sum += dis

        ans = min(ans, sum)
        return

    for i in range(idx, len(chicken)):
        ch.append(i)
        go(idx+1, cnt+1)
        ch.pop()


N, M = map(int, input().split())
arr = [[0] * N for _ in range(N)]
for i in range(N):
    arr[i] = list(map(int, input().split()))

for i in range(N):
    for j in range(N):
        if arr[i][j] == 1:
            home.append((i, j))
        elif arr[i][j] == 2:
            chicken.append((i, j))
go(0, 0)

print(ans)
